File: apps/dsa-web/src/api/watchlist.ts
import apiClient from './index';
import type {
  TagItem,
  TagCreate,
  TagUpdate,
  GroupItem,
  GroupCreate,
  GroupUpdate,
  StockAdd,
  StockListItem,
  StockListResponse,
  StockHistoryResponse,
  MessageResponse,
  // New types
  GroupInfo,
  ItemInfo,
  GroupListResp,
  ItemListResp,
  ItemSearchResp,
} from '../types/watchlist';

const BASE = '/api/v1/watchlist';

export const watchlistApi = {
  // ============ Group Operations (New API) ============

  listGroups: async (): Promise<GroupInfo[]> => {
    const response = await apiClient.get<GroupListResp>(`${BASE}/group/list`);
    return response.data.groups || [];
  },

  createGroup: async (name: string): Promise<void> => {
    await apiClient.post(`${BASE}/group/create`, { name });
  },

  updateGroup: async (id: number, name: string): Promise<void> => {
    await apiClient.put(`${BASE}/group/update`, { id, name });
  },

  deleteGroup: async (id: number): Promise<void> => {
    await apiClient.delete(`${BASE}/group/delete?id=${id}`);
  },

  sortGroups: async (items: number[]): Promise<void> => {
    await apiClient.put(`${BASE}/group/sort`, { items });
  },

  // ============ Item Operations (New API) ============

  listItems: async (groupId: number, size = 20, offset = 0): Promise<{ items: ItemInfo[]; total: number }> => {
    const response = await apiClient.get<ItemListResp>(
      `${BASE}/item/list?groupId=${groupId}&size=${size}&offset=${offset}`
    );
    return {
      items: response.data.items || [],
      total: response.data.total || 0,
    };
  },

  addItem: async (tsCode: string, groupIds: number[]): Promise<void> => {
    await apiClient.post(`${BASE}/item/add`, { tsCode, groupIds });
  },

  removeItem: async (tsCode: string, groupId: number): Promise<void> => {
    await apiClient.delete(`${BASE}/item/remove?tsCode=${tsCode}&groupId=${groupId}`);
  },

  moveItem: async (tsCode: string, fromGroupId: number, toGroupId: number): Promise<void> => {
    await apiClient.put(`${BASE}/item/move`, { tsCode, fromGroupId, toGroupId });
  },

  sortItems: async (groupId: number, items: { tsCode: string; action: string }[]): Promise<void> => {
    await apiClient.put(`${BASE}/item/sort`, { groupId, items });
  },

  searchStocks: async (keyword: string, limit = 10): Promise<ItemSearchResp> => {
    const response = await apiClient.get<ItemSearchResp>(
      `${BASE}/item/search?keyword=${encodeURIComponent(keyword)}&limit=${limit}`
    );
    return response.data;
  },

  // ============ Legacy Stock Operations ============

  getStocks: async (params?: {
    groupId?: number;
    tagId?: number;
    page?: number;
    limit?: number;
  }): Promise<StockListResponse> => {
    const query = new URLSearchParams();
    if (params?.groupId) query.set('groupId', String(params.groupId));
    if (params?.tagId) query.set('tagId', String(params.tagId));
    if (params?.page) query.set('page', String(params.page));
    if (params?.limit) query.set('limit', String(params.limit));
    const response = await apiClient.get<StockListResponse>(`${BASE}/stocks?${query}`);
    return response.data;
  },

  addStock: async (data: StockAdd): Promise<StockListItem> => {
    const response = await apiClient.post<StockListItem>(`${BASE}/stocks`, data);
    return response.data;
  },

  deleteStock: async (code: string): Promise<MessageResponse> => {
    const response = await apiClient.delete<MessageResponse>(`${BASE}/stocks/${code}`);
    return response.data;
  },

  getStockHistory: async (code: string, params?: { page?: number; limit?: number }): Promise<StockHistoryResponse> => {
    const query = new URLSearchParams();
    if (params?.page) query.set('page', String(params.page));
    if (params?.limit) query.set('limit', String(params.limit));
    const response = await apiClient.get<StockHistoryResponse>(`${BASE}/stocks/${code}/history?${query}`);
    return response.data;
  },

  setStockTags: async (code: string, tagIds: number[]): Promise<MessageResponse> => {
    const response = await apiClient.post<MessageResponse>(`${BASE}/stocks/${code}/tags`, { tag_ids: tagIds });
    return response.data;
  },

  setStockGroup: async (code: string, groupId: number | null): Promise<MessageResponse> => {
    const response = await apiClient.put<MessageResponse>(`${BASE}/stocks/${code}/group`, { group_id: groupId });
    return response.data;
  },

  // ============ Legacy Tag Operations ============

  getTags: async (): Promise<TagItem[]> => {
    const response = await apiClient.get<TagItem[]>(`${BASE}/tags`);
    return response.data;
  },

  createTag: async (data: TagCreate): Promise<TagItem> => {
    const response = await apiClient.post<TagItem>(`${BASE}/tags`, data);
    return response.data;
  },

  updateTag: async (id: number, data: TagUpdate): Promise<TagItem> => {
    const response = await apiClient.put<TagItem>(`${BASE}/tags/${id}`, data);
    return response.data;
  },

  deleteTag: async (id: number): Promise<MessageResponse> => {
    const response = await apiClient.delete<MessageResponse>(`${BASE}/tags/${id}`);
    return response.data;
  },

  // ============ Legacy Group Operations ============

  getGroups: async (): Promise<GroupItem[]> => {
    const response = await apiClient.get<GroupItem[]>(`${BASE}/groups`);
    return response.data;
  },

  createGroupLegacy: async (data: GroupCreate): Promise<GroupItem> => {
    const response = await apiClient.post<GroupItem>(`${BASE}/groups`, data);
    return response.data;
  },

  updateGroupLegacy: async (id: number, data: GroupUpdate): Promise<GroupItem> => {
    const response = await apiClient.put<GroupItem>(`${BASE}/groups/${id}`, data);
    return response.data;
  },

  deleteGroupLegacy: async (id: number): Promise<MessageResponse> => {
    const response = await apiClient.delete<MessageResponse>(`${BASE}/groups/${id}`);
    return response.data;
  },
};

File: api/v1/endpoints/watchlist.py
# -*- coding: utf-8 -*-
"""自选股 API 端点 - 升级版"""

import logging
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from api.deps import get_db
from api.v1.schemas.watchlist import (
    AccuracyStats,
    AnalysisHistoryItem,
    GroupCreate,
    GroupUpdate,
    GroupSort,
    GroupListResp,
    GroupInfo,
    ItemAdd,
    ItemMove,
    ItemSort,
    ItemListResp,
    ItemSearchResp,
    ItemSearchInfo,
    MessageResp,
    StockHistoryResp,
    TagInfo,
    TagItem,
    TagCreate,
    TagUpdate,
    StockTagSet,
    ItemInfo,
)
from src.services.watchlist_service import WatchlistService

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/watchlist", tags=["watchlist"])


# ========== 分组接口 ==========

@router.get("/group/list", response_model=GroupListResp, summary="获取分组列表")
def list_groups(db: Session = Depends(get_db)):
    """获取分组列表（含"全部"虚拟分组）"""
    service = WatchlistService(db)
    result = service.list_groups()
    return GroupListResp(groups=[GroupInfo(**g) for g in result['groups']])


@router.post("/group/create", response_model=GroupInfo, summary="创建分组")
def create_group(data: GroupCreate, db: Session = Depends(get_db)):
    """创建分组"""
    service = WatchlistService(db)
    try:
        result = service.create_group(data.name)
        return GroupInfo(id=result['id'], name=result['name'], sortOrder=0, stockCount=0, isDefault=False)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.put("/group/update", response_model=GroupInfo, summary="更新分组")
def update_group(data: GroupUpdate, db: Session = Depends(get_db)):
    """更新分组名称"""
    service = WatchlistService(db)
    result = service.update_group(data.id, data.name)
    if not result:
        raise HTTPException(status_code=404, detail="分组不存在")
    return GroupInfo(id=result['id'], name=result['name'], sortOrder=0, stockCount=0, isDefault=False)


@router.delete("/group/delete", response_model=MessageResp, summary="删除分组")
def delete_group(id: int = Query(..., description="分组ID"), db: Session = Depends(get_db)):
    """删除分组"""
    service = WatchlistService(db)
    if not service.delete_group(id):
        raise HTTPException(status_code=404, detail="分组不存在")
    return MessageResp(message="success")


@router.put("/group/sort", response_model=MessageResp, summary="分组排序")
def sort_groups(data: GroupSort, db: Session = Depends(get_db)):
    """设置分组排序顺序"""
    service = WatchlistService(db)
    service.sort_groups(data.items)
    return MessageResp(message="success")


# ========== 条目接口 ==========

@router.get("/item/list", response_model=ItemListResp, summary="获取条目列表")
def list_items(
    groupId: int = Query(0, description="分组ID，0表示全部"),
    size: int = Query(20, ge=1, le=100),
    offset: int = Query(0, ge=0),
    db: Session = Depends(get_db),
):
    """获取条目列表（含行情数据）"""
    service = WatchlistService(db)
    result = service.list_items(groupId, size, offset)
    items = [ItemInfo(**item) for item in result['items']]
    return ItemListResp(items=items, total=result['total'])


@router.post("/item/add", response_model=MessageResp, summary="添加条目")
def add_item(data: ItemAdd, db: Session = Depends(get_db)):
    """添加条目到分组"""
    service = WatchlistService(db)
    service.add_item(data.tsCode, data.groupIds)
    return MessageResp(message="success")


@router.delete("/item/remove", response_model=MessageResp, summary="删除条目")
def remove_item(
    tsCode: str = Query(..., description="股票代码"),
    groupId: int = Query(0, description="分组ID"),
    db: Session = Depends(get_db),
):
    """从指定分组删除条目"""
    service = WatchlistService(db)
    if not service.remove_item(tsCode, groupId):
        raise HTTPException(status_code=404, detail="条目不存在")
    return MessageResp(message="success")


@router.put("/item/move", response_model=MessageResp, summary="移动条目")
def move_item(data: ItemMove, db: Session = Depends(get_db)):
    """移动条目到其他分组"""
    service = WatchlistService(db)
    service.move_item(data.tsCode, data.fromGroupId, data.toGroupId)
    return MessageResp(message="success")


@router.put("/item/sort", response_model=MessageResp, summary="条目排序")
def sort_items(data: ItemSort, db: Session = Depends(get_db)):
    """条目排序（置顶/置底）"""
    service = WatchlistService(db)
    items = [{'tsCode': item.tsCode, 'action': item.action} for item in data.items]
    service.sort_items(data.groupId, items)
    return MessageResp(message="success")


@router.get("/item/search", response_model=ItemSearchResp, summary="搜索股票")
def search_stocks(
    keyword: str = Query(..., min_length=1, max_length=20),
    limit: int = Query(10, ge=1, le=50),
    db: Session = Depends(get_db),
):
    """搜索股票代码/名称"""
    service = WatchlistService(db)
    results = service.search_stocks(keyword, limit)
    items = [ItemSearchInfo(**r) for r in results]
    return ItemSearchResp(items=items)


# ========== 标签接口 ==========

@router.get("/tags", response_model=List[TagItem], summary="获取所有标签")
def list_tags(db: Session = Depends(get_db)):
    """获取用户的所有标签"""
    service = WatchlistService(db)
    result = service.list_tags()
    return [TagItem(**t) for t in result]


@router.post("/tags", response_model=TagItem, summary="创建标签")
def create_tag(data: TagCreate, db: Session = Depends(get_db)):
    """创建标签"""
    service = WatchlistService(db)
    try:
        result = service.create_tag(data.name, data.color or "#00d4ff")
        return TagItem(**result)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.put("/tags/{tag_id}", response_model=TagItem, summary="更新标签")
def update_tag(tag_id: int, data: TagUpdate, db: Session = Depends(get_db)):
    """更新标签"""
    service = WatchlistService(db)
    try:
        result = service.update_tag(tag_id, data.name, data.color)
        if not result:
            raise HTTPException(status_code=404, detail="标签不存在")
        return TagItem(**result)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.delete("/tags/{tag_id}", response_model=MessageResp, summary="删除标签")
def delete_tag(tag_id: int, db: Session = Depends(get_db)):
    """删除标签"""
    service = WatchlistService(db)
    if not service.delete_tag(tag_id):
        raise HTTPException(status_code=404, detail="标签不存在")
    return MessageResp(message="success")


@router.get("/stocks/{code}/tags", response_model=List[TagInfo], summary="获取股票标签")
def get_stock_tags(code: str, db: Session = Depends(get_db)):
    """获取某只股票的标签"""
    service = WatchlistService(db)
    return [TagInfo(id=t.id, name=t.name, color=t.color) for t in service.get_stock_tags(code)]


@router.post("/stocks/{code}/tags", response_model=MessageResp, summary="设置股票标签")
def set_stock_tags(code: str, data: StockTagSet, db: Session = Depends(get_db)):
    """设置股票的标签"""
    service = WatchlistService(db)
    try:
        service.set_stock_tags(code, data.tag_ids)
        return MessageResp(message="success")
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


# ========== 分析历史 ==========

@router.get("/stocks/{code}/history", response_model=StockHistoryResp, summary="获取分析历史")
def get_stock_analysis_history(
    code: str,
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
):
    """获取单只股票的历史分析记录和回测统计"""
    service = WatchlistService(db)
    result = service.get_stock_analysis_history(code, page, limit)
    return StockHistoryResp(
        items=[AnalysisHistoryItem(**item) for item in result['items']],
        total=result['total'],
        page=result['page'],
        limit=result['limit'],
        accuracyStats=AccuracyStats(**result['accuracyStats']) if result['accuracyStats'] else None,
    )
